### ISSUE TITLE
Generate correct looping limits for coloured loops

### ISSUE THREAD

**Issue by thomasmelvin** [stfc/PSyclone#127]
When computing the horizontal looping for coloured loops the correct infrastructure calls are not currently generated, these should be:
`do cell=1,mesh%get_last_halo_cell_per_colour_any(colour,1)` for continuous function spaces and 
`do cell=1,mesh%get_last_edge_cell_per_colour(colour)` for discontinuous function spaces.

These calls have been demonstrated in psy-lite in Lfric #1302 [https://code.metoffice.gov.uk/trac/lfric/ticket/1302](url)

**Comment by rupertford** [comment]
Hi Tom, to start off with could you confirm that we are currently generating the following ...

`do cell=1,mesh%get_last_halo_cell_per_colour(colour,1)` for continuous function spaces

and PSyclone should refuse to colour a discontinuous function space saying 'Loops iterating over a discontinuous function space are not currently supported'

That is what I would expect us to produce at the moment.

Thanks.

**Comment by rupertford** [comment]
This issue has been discussed with Tom by email.

1) The current PSyclone release (and the latest version of PSyclone on master) uses the old colouring API. The redundant colouring pull request #115 should address this problem when it is accepted on master and subsequently included in a release.

2) Tom also performs colouring for a discontinuous loop. This is not supported in the current #115 pull request and is the subject of a separate issue #116. However, colouring a discontinuous loop should not really be required and the only reason it is being done is that PSyclone treats wtheta as a continuous quantity. Therefore, sorting issue #84 out would fix the problem. I propose working on #84 next.

**Comment by rupertford** [comment]
I've just taken a look at #84 and Iva has taken it on. I've contacted her to ask the status.

**Comment by arporter** [comment]
@thomasmelvin, just to let you know that #115 is now on master so PSyclone should now be using the up-to-date colouring API. #84 is still in progress.

**Comment by TeranIvy** [comment]
Hi @arporter  and @thomasmelvin!

I can confirm that after making Wtheta discontinuous (I #84, PR #128) PSyclone does not attempt to colour loops over it any more. I can also confirm that, following changes in #115, PSyclone raises GenerationError when trying to colour loop over discontinuous field writers.

I added an example of `Wtheta` writer from the LFRic trunk to #128, which led to a couple of discoveries.

1) One of the algorithm subroutines had a commented invoke to a `Wtheta` writer in `sample_poly_adv_kernel_mod.F90` kernel as it attempted to colour a discontinuous loop. I enabled that code as `Wtheta `is now discontinuous and I am happy to report that the code generation works. This can be done in the LFRic original algorithm (`advection_alg_mod.x90`) as well after this support is available and the relevant PSyKAl-lite code can also be removed. Also, there was a bug in number of metadata in `sample_poly_adv_kernel_mod.F90` (unnecessary `arg_type(GH_FIELD,   GH_READ,  Wtheta`) which will have to be removed in the original LFRic `sample_poly_adv_kernel_mod.F90`.

2) Two LFRic kernels in the added example, `rtheta_wtheta_kernel_mod.F90` and `rtheta_bd_kernel_mod.F90`, have `GH_INC` access for `Wtheta` which should really be `GH_READWRITE`. As this support is being added in #25, I changed `GH_INC` Wtheta to `GH_WRITE` for now and left comments in the code.
However, a side effect of PR #128 is that a `GH_INC` access raises an error for discontinuous FS, so I will need to finish #25 immediately #128 is merged to be able to replace `GH_INC` instances of `Wtheta` and other discontinuous quantities (such as operators) with `GH_READWRITE` in the LFRic code.
